Fixes testing container layout issues for QUnit 2.14

drewlee opened the pull request:
This change attempts to address the numerous issues that have been introduced by the CSS positional updates in QUnit 2.14.0, including https://github.com/emberjs/ember-qunit/issues/793 and https://github.com/emberjs/ember-qunit/issues/801.  Besides visual layout problems, the relative percentage-based dimensions introduced in PR https://github.com/emberjs/ember-qunit/pull/786/ have been found to cause random test failures in existing applications, especially those which assert visibility states. The modifications did not consider that the ember testing container resides in qunit's fixture for v5. These have also been found to cause `ember-a11y-testing` failures with false-positives/negatives.

Given that the QUnit layout is now `position: fixed` based, it is no longer possible for the testing container to fall directly below the test results. As such, the container now also requires fixed positioning in relation to the viewport, similar to the "docked container" layout.

To follow the convention applied in https://github.com/qunitjs/qunit/pull/1513/files, the proposed fixes ensure that the UI reverts to the old layouts if the browser doesn't support CSS `flex` display, or the viewport height is constricted below the 500px threshold. Percentage based dimensions have been reverted back to fixed pixels to prevent testing container jank and false-positive/negative failures mentioned above. I've iterated over a bunch of different ways of remediating these problems, and believe this is probably the most balanced approach between usability and not having to override a bunch of QUnit's styling.

For the default layout, the proposed fix pins the testing container to the bottom of the viewport and centered, essentially dividing the screen into two horizontal panes with some semblance of the old layout.

Default view:
![image](https://user-images.githubusercontent.com/945868/107805308-037c7f00-6d1a-11eb-86b2-45e021df84fe.png)
Test results scrolled to bottom:
![image](https://user-images.githubusercontent.com/945868/107805477-43dbfd00-6d1a-11eb-8a3d-21117df16b76.png)

Viewport below 500px height (or lack of `flex` support) reverts back to old layout.
Top:
![image](https://user-images.githubusercontent.com/945868/107805982-01ff8680-6d1b-11eb-8faa-fb1b9a67cfb4.png)
Bottom:
![image](https://user-images.githubusercontent.com/945868/107805992-062ba400-6d1b-11eb-89a1-4c86522fc37b.png)

The "docked container" option pretty much remains similar and pins the testing container to the bottom right side of the screen.

Default view:
![image](https://user-images.githubusercontent.com/945868/107806211-615d9680-6d1b-11eb-87b5-a2e01a797861.png)

Test results scrolled to bottom:
![image](https://user-images.githubusercontent.com/945868/107806241-6ae6fe80-6d1b-11eb-9997-e5b56a50f31b.png)

Height constriction (or lack of `flex` support) reverts to old layout
![image](https://user-images.githubusercontent.com/945868/107806259-70444900-6d1b-11eb-9db2-fd4f1d45cacf.png)
![image](https://user-images.githubusercontent.com/945868/107806273-74706680-6d1b-11eb-995c-21b3913cc146.png)

Dev mode:
<img width="1272" alt="Screen Shot 2021-02-12 at 11 24 24 AM" src="https://user-images.githubusercontent.com/945868/107813024-00d35700-6d25-11eb-821b-0516fa768cfb.png">

Notable items:
- Since the percentage-based dimensions have been removed, the fixes applied in PR https://github.com/emberjs/ember-qunit/pull/804/ are no longer necessary as "full screen" mode works as expected.
- These changes require a new qualifier to differentiate between standard layout and "docked container" layout, hence the addition of the `docked-container` CSS class to the `body`.
- Theoretically, these changes are backportable to v.4.0.x.
- To simplify things, we may want to consider making the "docked container" layout the new default and remove it as a selectable option from the UI altogether.
- These changes still require some testing in various browser permutations, and also verification in real applications. But, I think we're in a good enough state to open it up for discussion.

drewlee commented:
Marking this as a draft for cross-browser and app testing.

scalvert commented:
Wow thanks for doing this, @drewlee. Really appreciate you looking into fixing this.

drewlee commented:
Updated PR with the following changes:
- Removed scripted `relative` position styling in favor of a CSS default. This makes it easier to override with `fixed` positioning for the new layout without having to resort to using `!important` declarations.
- Removed scripted `marginBottom` styling in favor of a CSS default. Again, this makes it easier to override without resorting to `!important` declarations.
- Added scripted `marginBottom` override when the container is hidden. Otherwise, the full page isn't used to show test results and looks odd.
- Updated styling for bottom margin on test results for docked container layout. The initial implementation didn't work with Firefox on Win/Mac.

Visually verified fixes on Safari, Chrome (Win/Mac), Firefox (Win/Mac), and Edge (Win/Mac).
Ran fixes in several large-scale apps. All related tests pass as expected.

Example layout with hidden container:
<img width="1190" alt="Screen Shot 2021-02-16 at 5 21 34 PM" src="https://user-images.githubusercontent.com/945868/108134318-a0714c00-707b-11eb-9597-a48900d9a4e2.png">

drewlee commented:
I created a minimal reproduction case demonstrating the visibility-related false-negatives at https://github.com/drewlee/ember-qunit-layout-bugs/. Yarn linking to this set of fixes allows the test to pass as expected.

rwjblue commented:
> To simplify things, we may want to consider making the "docked container" layout the new default and remove it as a selectable option from the UI altogether.

Agree. How much more work is this to include here?

scalvert reviewed (approved):
This looks really good to me. Once @rwjblue's question is sorted I think this is GTG.

drewlee commented:
@rwjblue should be fairly straightforward, would just need to remove and streamline some of the CSS/JS. I can probably push out an update by EOD.

drewlee commented:
Makes "docked container" layout the default, which effectively means that:
1. The testing container is always in `fixed` position, so it's no longer necessary to modify position property via script
2. The "Dock container" option is removed from the UI
3. Bottom margin for test results set in the CSS instead of via scripting
4. The overall implementation is greatly simplified as scripted checks are no longer necessary to conditionally apply styling based on layout type

I've found it necessary to apply a small right gutter for the testing container so that it doesn't occlude the test results scroll bar. This is done automatically by the browser when the layout reverts to the previous version (below 500px height).

Scrolled top:
<img width="1203" alt="Screen Shot 2021-02-17 at 6 06 55 PM" src="https://user-images.githubusercontent.com/945868/108284538-07116b00-714b-11eb-87cf-d2a63a65925a.png">
Scrolled bottom:
<img width="1203" alt="Screen Shot 2021-02-17 at 6 06 59 PM" src="https://user-images.githubusercontent.com/945868/108284548-0bd61f00-714b-11eb-9b5b-5b5e0719bb54.png">
Under 500px height:
<img width="1203" alt="Screen Shot 2021-02-17 at 6 07 14 PM" src="https://user-images.githubusercontent.com/945868/108284557-12649680-714b-11eb-9136-fbee57a4f4a5.png">
